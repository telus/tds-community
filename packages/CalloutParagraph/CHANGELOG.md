# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.6.0](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.5.1...@tds/community-callout-paragraph@0.6.0) (2021-09-21)


### Features

* **community-callout-paragraph:** react 17 and styled components 5 ([e9cfca7](https://github.com/telus/tds-community/commit/e9cfca7a833f52b90decd2bc3646d886741fcf98))





## [0.5.1](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.5.0...@tds/community-callout-paragraph@0.5.1) (2021-03-02)


### Bug Fixes

* **community-callout-paragraph:** increase icon margin ([e0cc1cb](https://github.com/telus/tds-community/commit/e0cc1cb4684feb9865dc3956104177f74cdec277))





# [0.5.0](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.4.0...@tds/community-callout-paragraph@0.5.0) (2021-03-02)


### Features

* **community-callout-paragraph:** add textAlign prop ([43cfb72](https://github.com/telus/tds-community/commit/43cfb72bfd2e05db43d1ba4a1a1b59a32738d931))





# [0.4.0](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.3.1...@tds/community-callout-paragraph@0.4.0) (2021-03-01)


### Features

* **community-callout-paragraph:** add icon support ([797e6e1](https://github.com/telus/tds-community/commit/797e6e1dd0f0dd7dafa18059086ca1be11d50489))





## [0.3.1](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.3.0...@tds/community-callout-paragraph@0.3.1) (2020-11-19)


### Bug Fixes

* **community-callout-paragraph:** fix margins for mobile width ([17f0155](https://github.com/telus/tds-community/commit/17f01559fceb9102a05fe35829875b70d8427bb1))





# [0.3.0](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.2.2...@tds/community-callout-paragraph@0.3.0) (2020-11-17)


### Features

* **community-callout-paragraph:** add compact prop ([4e7105a](https://github.com/telus/tds-community/commit/4e7105a6d6e5fe8325694d077ff9cbbb2b261c7f))





## [0.2.2](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.2.1...@tds/community-callout-paragraph@0.2.2) (2020-11-10)


### Bug Fixes

* **community-callout-paragraph:** reduce font size and add line height ([94f3867](https://github.com/telus/tds-community/commit/94f3867505ec45b675a2237aff6087d169bc6b56))





## [0.2.1](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.2.0...@tds/community-callout-paragraph@0.2.1) (2020-11-05)


### Bug Fixes

* **community-callout-paragraph:** reduce padding amount to 2px ([50ca864](https://github.com/telus/tds-community/commit/50ca864771cbf568233bef73fe89635fa5343b8a))
* **community-callout-paragraph:** reduce padding to match designs ([c7b8c91](https://github.com/telus/tds-community/commit/c7b8c91bde450f3c82652e897974705d290787e7))





# [0.2.0](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.1.1...@tds/community-callout-paragraph@0.2.0) (2020-11-04)


### Features

* **community-callout-paragraph:** add rounded corner option ([99018af](https://github.com/telus/tds-community/commit/99018afef99d1ac23b6af5e9cee2269fa6ff3037))





## [0.1.1](https://github.com/telus/tds-community/compare/@tds/community-callout-paragraph@0.1.0...@tds/community-callout-paragraph@0.1.1) (2020-09-09)


### Bug Fixes

* **community-callout-paragraph:** update util-helpers version ([8767507](https://github.com/telus/tds-community/commit/8767507941b0fe8b14bc01ef9d8025068fa8da81))





# 0.1.0 (2020-07-21)


### Bug Fixes

* **community-callout-paragraph:** setting correct package version ([03b2b15](https://github.com/telus/tds-community/commit/03b2b157d734a5f6cb1bd87c520dad68453872ef))


### Features

* **community-callout-paragraph:** create new component ([002f0ac](https://github.com/telus/tds-community/commit/002f0acaae1b51e8419d01173730d799a6a876f1))
