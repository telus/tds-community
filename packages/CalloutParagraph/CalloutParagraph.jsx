import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { safeRest } from '@tds/util-helpers'
import { componentWithName } from '@tds/util-prop-types'

import Paragraph from '@tds/core-paragraph'
import { colorTelusPurple } from '@tds/core-colours'

const marginValue = {
  mobile: {
    default: '-1.5rem',
    intermediate: '-1.5rem',
    narrow: '-1rem',
    compact: '-1rem',
  },
  desktop: {
    default: '-2rem',
    intermediate: '-2rem',
    narrow: '-1rem',
    compact: '-1rem',
  },
}
const paddingValue = {
  mobile: {
    default: '1.5rem',
    intermediate: '1.5rem',
    narrow: '1rem',
    compact: '1rem',
  },
  desktop: {
    default: '2rem',
    intermediate: '2rem',
    narrow: '1rem',
    compact: '1rem',
  },
}

const TextWrapper = styled.div`
  margin-left: ${props => (props.roundedCorners ? 'auto' : marginValue.desktop[props.spacing])};
  margin-right: ${props => (props.roundedCorners ? 'auto' : marginValue.desktop[props.spacing])};
  @media (max-width: 576px) {
    margin-left: ${props => (props.roundedCorners ? 'auto' : marginValue.mobile[props.spacing])};
    margin-right: ${props => (props.roundedCorners ? 'auto' : marginValue.mobile[props.spacing])};
  }
  width: ${props => (props.roundedCorners ? '100%' : undefined)};
  p {
    position: relative;
    text-align: ${props => props.textAlign};
    font-size: ${props => (props.compact ? '14px' : '1rem')};
    line-height: ${props => (props.compact ? '20px' : undefined)};
    border-radius: ${props => (props.roundedCorners ? '5px' : undefined)};
    color: ${colorTelusPurple};
    background: #f2eff4;
    padding-top: ${props => (props.compact ? '2px' : '0.5rem')};
    padding-bottom: ${props => (props.compact ? '2px' : '0.5rem')};
    padding-left: ${props => paddingValue.desktop[props.spacing]};
    padding-right: ${props => paddingValue.desktop[props.spacing]};
    @media (max-width: 576px) {
      padding-left: ${props => paddingValue.mobile[props.spacing]};
      padding-right: ${props => paddingValue.mobile[props.spacing]};
    }
    > i {
      margin-right: 1rem;
      position: absolute;
      left: 10px;
    }
    > span {
      display: inline-block;
      margin-left: ${props => props.margin};
    }
  }
`

const iconMargin = (spacingLevel, hasIcon) => {
  if (hasIcon) {
    if (spacingLevel === 'compact') {
      return '1.75rem'
    }
    return '0.75rem'
  }
  return undefined
}

// Format text align to be backwards compatible.
const formatTextAlign = (textAlignProp, hasRoundedCorners) => {
  if (textAlignProp) {
    return textAlignProp
  } else if (hasRoundedCorners) {
    return 'center'
  }
  return 'left'
}

/**
 * @version ./package.json
 * @visibleName CalloutParagraph (beta)
 */
const CalloutParagraph = ({
  children,
  spacing,
  roundedCorners,
  compact,
  icon: Icon,
  textAlign,
  ...rest
}) => {
  return (
    <TextWrapper
      spacing={spacing}
      roundedCorners={roundedCorners}
      compact={compact}
      margin={iconMargin(spacing, Icon)}
      textAlign={formatTextAlign(textAlign, roundedCorners)}
    >
      <Paragraph {...safeRest(rest)}>
        {Icon && <Icon />}
        <span>{children}</span>
      </Paragraph>
    </TextWrapper>
  )
}

CalloutParagraph.propTypes = {
  /**
   *
   */
  spacing: PropTypes.oneOf(['default', 'narrow', 'compact', 'intermediate']),

  /**
   * The content. Can be text, other components, or HTML elements.
   */
  children: PropTypes.node.isRequired,
  /**
   * Corners will be slightly rounded, and text will be center-aligned when passed.
   */
  roundedCorners: PropTypes.bool,
  /**
   * Font size and padding around text will be smaller.
   */
  compact: PropTypes.bool,
  /**
   * Provide an icon from the Dependent icon group in `@tds/core-interactive-icon`.
   */
  icon: componentWithName('Decorative', true),
  /**
   * Set explicit text-align property. If not set, default text-align is 'center' for
   * `roundedCorners` and 'left' otherwise.
   */
  textAlign: PropTypes.oneOf(['center', 'left', 'right']),
}

CalloutParagraph.defaultProps = {
  spacing: 'default',
  roundedCorners: false,
  compact: false,
  icon: undefined,
  textAlign: undefined,
}

export default CalloutParagraph
