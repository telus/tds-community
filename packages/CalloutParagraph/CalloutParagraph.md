### Usage criteria

- Use `spacing` prop according to tds-core-card component
- Use all props of tds-core-paragraph component

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={5}>
      <Card variant="defaultWithBorder">
        <Box between={3}>
          <Heading level="h3">Internet 15 - Special Offer</Heading>
          <CalloutParagraph>
            Our commitment to service is demonstrated in everything we do.
          </CalloutParagraph>
          <HairlineDivider />
          <Text size="medium">Good for basic browsing, and posting to social media.</Text>
          <Button>Add to cart</Button>
          <ChevronLink href="#">Learn more</ChevronLink>
        </Box>
      </Card>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Rounded Corners

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={5}>
      <Card variant="defaultWithBorder">
        <Box between={3}>
          <Heading level="h3">Internet 15 - Special Offer</Heading>
          <CalloutParagraph roundedCorners>Save $70/month.</CalloutParagraph>
          <HairlineDivider />
          <Text size="medium">Good for basic browsing, and posting to social media.</Text>
          <Button>Add to cart</Button>
          <ChevronLink href="#">Learn more</ChevronLink>
        </Box>
      </Card>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Icons

- Icon will be displayed before text
- Must be valid TDS SVGIcon name. Refer to TDS documentation for full list

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={7} lg={6}>
      <FlexGrid.Row>
        <FlexGrid.Col xl={11}>
          <CalloutParagraph bold roundedCorners textAlign="left" icon={Deals}>
            Order online for a $300 bill credit.
          </CalloutParagraph>
        </FlexGrid.Col>
      </FlexGrid.Row>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```

### Compact

```jsx
<FlexGrid>
  <FlexGrid.Row>
    <FlexGrid.Col xs={12} md={5}>
      <Card variant="defaultWithBorder">
        <Box between={3}>
          <Heading level="h3">Internet 15 - Special Offer</Heading>
          <CalloutParagraph roundedCorners compact>
            Save $70/month.
          </CalloutParagraph>
          <HairlineDivider />
          <Text size="medium">Good for basic browsing, and posting to social media.</Text>
          <Button>Add to cart</Button>
          <ChevronLink href="#">Learn more</ChevronLink>
        </Box>
      </Card>
    </FlexGrid.Col>
  </FlexGrid.Row>
</FlexGrid>
```
