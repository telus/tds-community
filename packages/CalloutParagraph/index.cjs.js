const CalloutParagraph = require('./dist/index.cjs')

module.exports = CalloutParagraph
