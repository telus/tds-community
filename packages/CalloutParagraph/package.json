{
  "name": "@tds/community-callout-paragraph",
  "version": "0.6.0",
  "description": "",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telus/tds-community.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telus/tds-community/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2 || ^17.0.0",
    "react-dom": "^16.8.2 || ^17.0.0",
    "styled-components": "^4.1.3 || ^5.0.0"
  },
  "dependencies": {
    "@tds/core-colours": "^2.2.1",
    "@tds/core-paragraph": "^2.0.12",
    "@tds/util-helpers": "^1.5.0",
    "prop-types": "^15.6.2"
  }
}
