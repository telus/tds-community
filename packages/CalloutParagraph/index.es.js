import CalloutParagraph from './dist/index.es'

export default CalloutParagraph
