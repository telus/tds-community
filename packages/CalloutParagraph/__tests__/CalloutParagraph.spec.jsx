import React from 'react'
import { shallow } from 'enzyme'
import { Deals } from '@tds/core-decorative-icon'

import CalloutParagraph from '../CalloutParagraph'

describe('CalloutParagraph', () => {
  const defaultProps = { children: 'Text' }

  const doShallow = props => shallow(<CalloutParagraph {...defaultProps} {...props} />)

  it('renders', () => {
    const calloutParagraph = doShallow()

    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph).toHaveProp('spacing', 'default')
  })

  it('renders with narrow', () => {
    const calloutParagraph = doShallow({ spacing: 'narrow' })
    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph).toHaveProp('spacing', 'narrow')
  })

  it('renders with compact', () => {
    const calloutParagraph = doShallow({ spacing: 'compact' })
    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph).toHaveProp('spacing', 'compact')
  })

  it('renders with intermediate', () => {
    const calloutParagraph = doShallow({ spacing: 'intermediate' })
    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph).toHaveProp('spacing', 'intermediate')
  })

  it('renders an icon', () => {
    const calloutParagraph = doShallow({ icon: Deals })
    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph.find('DecorativeIcon')).toExist()
  })

  it('specifies text align', () => {
    const calloutParagraph = doShallow({ textAlign: 'center' })
    expect(calloutParagraph).toMatchSnapshot()
    expect(calloutParagraph).toHaveProp('textAlign', 'center')
  })
})
