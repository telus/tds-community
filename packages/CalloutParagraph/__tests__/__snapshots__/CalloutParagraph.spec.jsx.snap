// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalloutParagraph renders 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  roundedCorners={false}
  spacing="default"
  textAlign="left"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;

exports[`CalloutParagraph renders an icon 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  margin="0.75rem"
  roundedCorners={false}
  spacing="default"
  textAlign="left"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <DecorativeIcon />
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;

exports[`CalloutParagraph renders with compact 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  roundedCorners={false}
  spacing="compact"
  textAlign="left"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;

exports[`CalloutParagraph renders with intermediate 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  roundedCorners={false}
  spacing="intermediate"
  textAlign="left"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;

exports[`CalloutParagraph renders with narrow 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  roundedCorners={false}
  spacing="narrow"
  textAlign="left"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;

exports[`CalloutParagraph specifies text align 1`] = `
<CalloutParagraph__TextWrapper
  compact={false}
  roundedCorners={false}
  spacing="default"
  textAlign="center"
>
  <Paragraph
    align="left"
    bold={false}
    invert={false}
    size="medium"
  >
    <span>
      Text
    </span>
  </Paragraph>
</CalloutParagraph__TextWrapper>
`;
